File: solutions/python/sublist/1/sublist.py
"""
This exercise stub and the test suite contain several enumerated constants.

Enumerated constants can be done with a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it's memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""

# Possible sublist categories.
# Change the values as you see fit.
SUBLIST = 1
SUPERLIST = 2
EQUAL = 3
UNEQUAL = 4


def sublist(list_one, list_two):
    """Determine the relationship between two lists."""
    # Helper function to check if list_a is contained in list_b
    def is_sublist(list_a, list_b):
        len_a = len(list_a)
        len_b = len(list_b)
        
        # An empty list is always a sublist
        if len_a == 0:
            return True
        
        # If list_a is longer than list_b, it can't be a sublist
        if len_a > len_b:
            return False
        
        # Check each possible starting position in list_b
        for i in range(len_b - len_a + 1):
            # Check if the slice of list_b matches list_a
            if list_b[i:i+len_a] == list_a:
                return True
        return False
    
    # Check for equality first (most specific case)
    if list_one == list_two:
        return EQUAL
    
    # Check if list_one is a sublist of list_two
    if is_sublist(list_one, list_two):
        return SUBLIST
    
    # Check if list_one is a superlist of list_two
    if is_sublist(list_two, list_one):
        return SUPERLIST
    
    # None of the above
    return UNEQUAL

File: solutions/c/sublist/1/sublist.c
#include "sublist.h"

// Helper function to check if 'small' is a contiguous sublist of 'big'
static int is_sublist(const int *small, size_t small_len,
                      const int *big,   size_t big_len) {
    if (small_len == 0) return 1;               // empty list is always a sublist
    if (big_len < small_len) return 0;          // not enough elements

    for (size_t i = 0; i <= big_len - small_len; ++i) {
        int match = 1;
        for (size_t j = 0; j < small_len; ++j) {
            if (big[i + j] != small[j]) {
                match = 0;
                break;
            }
        }
        if (match) return 1;
    }
    return 0;
}

comparison_result_t check_lists(int *list_to_compare, int *base_list,
                                size_t list_to_compare_element_count,
                                size_t base_list_element_count) {
    // Both empty -> equal
    if (list_to_compare_element_count == 0 && base_list_element_count == 0)
        return EQUAL;

    // One empty, the other not
    if (list_to_compare_element_count == 0)
        return SUBLIST;
    if (base_list_element_count == 0)
        return SUPERLIST;

    // Same length: check element-wise equality
    if (list_to_compare_element_count == base_list_element_count) {
        for (size_t i = 0; i < list_to_compare_element_count; ++i) {
            if (list_to_compare[i] != base_list[i])
                return UNEQUAL;
        }
        return EQUAL;
    }

    // Different lengths: one may be a sublist of the other
    if (list_to_compare_element_count < base_list_element_count) {
        if (is_sublist(list_to_compare, list_to_compare_element_count,
                       base_list, base_list_element_count))
            return SUBLIST;
    } else {
        if (is_sublist(base_list, base_list_element_count,
                       list_to_compare, list_to_compare_element_count))
            return SUPERLIST;
    }

    return UNEQUAL;
}